core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/AwsCrtVS34aHttpProperties.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.crt.internal.signer;
+
+import static software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner.CHUNKED_ENCODING;
+import static software.amazon.awssdk.http.auth.aws.AwsV4HttpSigner.PAYLOAD_SIGNING;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.safeParseLong;
+import static software.amazon.awssdk.http.auth.aws.util.SignerUtils.validatedProperty;
+
+import java.time.Clock;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.http.Header;
+import software.amazon.awssdk.http.auth.aws.checksum.ChecksumAlgorithm;
+import software.amazon.awssdk.http.auth.aws.util.CredentialScope;
+import software.amazon.awssdk.http.auth.aws.util.HttpChecksumUtils;
+import software.amazon.awssdk.http.auth.spi.SignRequest;
+import software.amazon.awssdk.http.auth.spi.SignerProperty;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+
+
+/**
+ * An interface which contains "properties" used by a crt-based v4a-signer at various steps.
+ * These properties are derived from the {@link SignerProperty}'s on a {@link SignRequest}.
+ */
+@SdkInternalApi
+public interface AwsCrtVS34aHttpProperties extends AwsCrtV4aHttpProperties {
+
+    static AwsCrtVS34aHttpProperties create(SignRequest<?, ? extends AwsCredentialsIdentity> signRequest) {
+        boolean payloadSigning = validatedProperty(signRequest, PAYLOAD_SIGNING, false);
+        boolean chunkedEncoding = validatedProperty(signRequest, CHUNKED_ENCODING, false);
+        boolean unsignedStreamingTrailer = signRequest.request().firstMatchingHeader("x-amz-content-sha256")
+            .map("STREAMING-UNSIGNED-PAYLOAD-TRAILER"::equals)
+            .orElse(false);
+        Optional<Long> maybeContentLength = safeParseLong(signRequest.request().firstMatchingHeader(Header.CONTENT_LENGTH)
+            .orElse(null));
+        Long contentLength = maybeContentLength
+            .orElse(signRequest.payload()
+                .map(HttpChecksumUtils::calculateContentLength)
+                .orElse(0L));
```

Comment:
If on line 50, contentLength is always non-null, it should be `long contentLength`.
But, unsure of multiple things here.
* safeParseLong ignores NumberFormatException - seems unsafe. If a Content-Length header is present, but not formatted correctly, it results in contentLength = 0, rather than failing.
* If payload is null, this results in contentLength = 0. not sure if that's fine v/s using `Long getContentLength()` or `Optional<Long> getContentLength`.

Reply:
Fair. I've changed it to throw instead of making an assumption. WRT the payload, if it's null or not present (it's an optional type), 0 is the expected content-length.